examples/platform/esp32/mode-support/static-supported-modes-manager.cpp

```diff
@@ -0,0 +1,173 @@
+                                ModeOptionsProvider(nullptr, nullptr));
+
+            memset(keyBuf, 0, sizeof(char) * ESP32Config::kMaxConfigKeyNameLength);
+            VerifyOrReturnValue(ESP32Config::KeyAllocator::SemanticTagMfgCode(keyBuf, sizeof(keyBuf), endpointId, index, stIndex) ==
+                                    CHIP_NO_ERROR,
+                                ModeOptionsProvider(nullptr, nullptr));
+            ESP32Config::Key stMfgCodeKey(ESP32Config::kConfigNamespace_ChipFactory, keyBuf);
+            VerifyOrReturnValue(ESP32Config::ReadConfigValue(stMfgCodeKey, semanticTagMfgCode) == CHIP_NO_ERROR,
+                                ModeOptionsProvider(nullptr, nullptr));
+
+            tag.value             = static_cast<uint16_t>(semanticTagValue);
+            tag.mfgCode           = static_cast<chip::VendorId>(semanticTagMfgCode);
+            semanticTags[stIndex] = static_cast<const SemanticTag>(tag);
+        }
+
+        option.label        = chip::CharSpan::fromCharString(modeLabelList[index].supportedModeLabel);
+        option.mode         = static_cast<uint8_t>(supportedModeMode);
+        option.semanticTags = DataModel::List<const SemanticTag>(semanticTags, semanticTagCount);
+
+        modeOptionStruct[index] = option;
+    }
+	endpointArray[endpointId][0] = static_cast<void *>(modeOptionStruct); 
+	endpointArray[endpointId][1] = static_cast<void *>(modeOptionStruct + supportedModeCount); 
+
+    return ModeOptionsProvider(modeOptionStruct, modeOptionStruct + supportedModeCount);
+}
+
+Status StaticSupportedModesManager::getModeOptionByMode(unsigned short endpointId, unsigned char mode,
+                                                        const ModeOptionStructType ** dataPtr) const
+{
+    auto modeOptionsProvider = this->getModeOptionsProvider(endpointId);
+    if (modeOptionsProvider.begin() == nullptr)
+    {
+        return Status::UnsupportedCluster;
+    }
+    auto * begin = modeOptionsProvider.begin();
+    auto * end   = modeOptionsProvider.end();
+
+    for (auto * it = begin; it != end; ++it)
+    {
+        auto & modeOption = *it;
+        if (modeOption.mode == mode)
+        {
+            *dataPtr = &modeOption;
+            return Status::Success;
+        }
+    }
+    emberAfPrintln(EMBER_AF_PRINT_DEBUG, "Cannot find the mode %u", mode);
+    return Status::InvalidCommand;
+}
+
+const ModeSelect::SupportedModesManager * ModeSelect::getSupportedModesManager()
+{
+    return &StaticSupportedModesManager::instance;
+}
+void StaticSupportedModesManager::FreeSupportedModes()
+{
+	for(int i=0; i<3; i++) {
+		if(endpointArray[i][0] != nullptr) {
+    		auto * begin = static_cast<ModeOptionStructType *>(endpointArray[i][0]);
+    		auto * end   = static_cast<ModeOptionStructType *>(endpointArray[i][1]);
+			for(auto *it = begin; it != end; ++it) {
+        		auto & modeOption = *it;
+				delete[] &modeOption.label;
+				delete[] &modeOption.semanticTags;
```

Comment:
This does not look right.  `modeOption.label` is a Span  We're taking the address of that span and then using array delete on it?  Similar for modeOption.semanticTags.

Was this tested?  Especially with something like asan enabled?

Reply:
Updated and tested memory leaks